internal/notifier/pagerduty.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	"k8s.io/utils/strings/slices"
+)
+
+var (
+	validEndpoints = []string{
+		"events.pagerduty.com",
+		"events.eu.pagerduty.com",
+	}
+)
+
+type PagerDuty struct {
+	Client     *pagerduty.Client
+	RoutingKey string
+}
+
+func NewPagerDuty(endpoint string, routingKey string) (*PagerDuty, error) {
+	pURL, err := url.ParseRequestURI(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("invalid PagerDuty endpoint URL %s: '%w'", endpoint, err)
+	}
+	pHostname := pURL.Hostname()
+	if !slices.Contains(validEndpoints, pHostname) {
+		return nil, fmt.Errorf("invalid PagerDuty endpoint hostname %s: should be one of: %v", pHostname, validEndpoints)
+	}
+	opts := []pagerduty.ClientOptions{
+		pagerduty.WithV2EventsAPIEndpoint("https://" + pHostname),
+	}
+	// No token needed for our usage of the V2 events API
```

Comment:
Not sure I undestand how this works, anyone can spam PagerDuty if it knows the URL?

Reply:
Yeah, the url is the same for everyone (except EU customers). The routing is done by the integration key. You can be rate limited, which I haven't handled yet.